File: frontend/src/types/contracts.ts
import type { ColumnTone } from './stage'

/** 계약 진행 상태. 매출 실적으로 잡히는 것은 '확정' 뿐입니다. */
export type ContractStatus = '확정' | '진행중' | '취소'

export type PipelineOutcomeCode = 'in_progress' | 'confirmed' | 'cancelled'

export type ContractKind = '신규 도입' | '증설' | '갱신' | '유지보수' | '소모품 공급'

/** API/DB에 저장하는 계약 유형 코드. 화면에서는 ContractKind의 한국어 라벨을 씁니다. */
export type ContractKindCode =
  'new_installation' | 'expansion' | 'renewal' | 'maintenance' | 'consumables_supply'

export interface ContractSeed {
  no: string
  /** 고객사. customers.ts 의 org 와 같은 표기를 씁니다. */
  org: string
  product: string
  amount: number
  kind: ContractKind
  status: ContractStatus
  /**
   * 계약일. 오늘로부터 며칠이며 과거이므로 음수입니다.
   * 아직 체결하지 않은 '진행중' 건은 협의를 시작한 날입니다.
   */
  signedOff: number
  /** 담당 영업 */
  owner: string
}

/** 실제 날짜와 지역이 붙은 계약 */
export interface Contract extends ContractSeed {
  /** 계약일 YYYY-MM-DD */
  date: string
  /** 고객사가 있는 시. regions.ts 에서 파생합니다. */
  region: string
  /** 화면에서 적어 넣은 메모. 시드에는 없습니다. */
  memo?: string
}

/** 계약 하나가 놓인 단계. 영업현황·계약현황이 저마다 다른 단계 집합을 씁니다. */
export interface StagedContract extends Contract {
  stageId: string
}

/** 계약 입력 폼이 내놓는 값. 계약번호·단계는 폼 밖에서 정합니다. */
export interface ContractDraft {
  org: string
  product: string
  amount: number
  kind: ContractKind
  owner: string
  /** YYYY-MM-DD */
  date: string
  memo: string
}

/** GET /api/pipeline-stages 응답. 순서는 position 오름차순입니다. */
export interface PipelineStageResponse {
  id: string
  name: string
  tone: ColumnTone
  outcome_code: PipelineOutcomeCode
  position: number
}

/** GET /api/products 응답. */
export interface ProductResponse {
  id: string
  name: string
  active: boolean
}

export interface ContractCreateRequest {
  customer_company_id: string
  contact_id: null
  product_id: string
  stage_id: string
  contract_type: ContractKindCode
  amount: number
  contract_date: string
  memo: string | null
}

export interface ContractPatchRequest {
  customer_company_id: string
  contact_id?: null
  product_id: string
  contract_type: ContractKindCode
  amount: number
  contract_date: string
  memo: string | null
}

export interface ContractMoveRequest {
  expected_stage_id: string
  stage_id: string
  position: number
}

/** GET /api/contracts 및 상세 조회 응답. 내부 UUID와 표시값을 함께 보존합니다. */
export interface ContractResponse {
  id: string
  contract_no: string
  customer_company_id: string
  customer_company_name: string
  customer_company_region_code: string | null
  contact_id: string | null
  contact_name: string | null
  owner_member_id: string
  owner_display_name: string
  product_id: string | null
  product_name: string | null
  stage_id: string
  stage_name: string
  stage_tone: ColumnTone
  stage_outcome_code: PipelineOutcomeCode
  stage_position: number
  title: string
  description: string | null
  contract_type: ContractKindCode
  amount: number
  contract_date: string
  ends_on: string | null
  warranty_terms: string | null
  expected_delivery_at: string | null
  memo: string | null
  position: number
  created_at: string
  updated_at: string
}

File: backend/app/schemas/contracts.py
from datetime import UTC, date, datetime, timedelta
from typing import Annotated, Literal, Self
from uuid import UUID

from pydantic import (
    AfterValidator,
    AwareDatetime,
    BaseModel,
    ConfigDict,
    Field,
    StrictInt,
    StringConstraints,
    model_validator,
)


def _seoul_offset(value: datetime) -> datetime:
    if value.utcoffset() != timedelta(hours=9):
        raise ValueError("datetime offset must be +09:00")
    return value


Text = Annotated[
    str,
    StringConstraints(strip_whitespace=True, strict=True, min_length=1, max_length=254),
]
LongText = Annotated[
    str,
    StringConstraints(strip_whitespace=True, strict=True, min_length=1, max_length=5_000),
]
SearchQuery = Annotated[
    str,
    StringConstraints(strip_whitespace=True, strict=True, min_length=1, max_length=100),
]
Money = Annotated[StrictInt, Field(ge=0, le=9_223_372_036_854_775_807)]
Position = Annotated[StrictInt, Field(ge=0, le=2_147_483_647)]
SafeDateTime = Annotated[
    AwareDatetime,
    Field(
        ge=datetime.min.replace(tzinfo=UTC),
        le=datetime.max.replace(hour=14, tzinfo=UTC),
    ),
    AfterValidator(_seoul_offset),
]
ContractType = Literal[
    "new_installation",
    "expansion",
    "renewal",
    "maintenance",
    "consumables_supply",
]
StageTone = Literal["gray", "blue", "purple", "orange", "green", "red"]
StageOutcome = Literal["in_progress", "confirmed", "cancelled"]


class _WriteModel(BaseModel):
    model_config = ConfigDict(extra="forbid")


class PipelineStageRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    name: str
    tone: StageTone
    outcome_code: StageOutcome
    position: int


class ProductRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    name: str
    active: bool


class ProductPage(BaseModel):
    items: list[ProductRead]
    skip: int
    limit: int
    total: int
    has_more: bool
    next_skip: int | None


class ProductPageParams(BaseModel):
    model_config = ConfigDict(extra="forbid")

    q: SearchQuery | None = None
    skip: int = Field(default=0, ge=0, le=9_223_372_036_854_775_807)
    limit: int = Field(default=30, ge=1, le=100)


class PipelineStageParams(BaseModel):
    model_config = ConfigDict(extra="forbid")


class ContractCreate(_WriteModel):
    customer_company_id: UUID
    contact_id: UUID | None = None
    product_id: UUID
    stage_id: UUID
    title: Text | None = None
    description: LongText | None = None
    contract_type: ContractType
    amount: Money
    contract_date: date
    ends_on: date | None = None
    warranty_terms: LongText | None = None
    expected_delivery_at: SafeDateTime | None = None
    memo: LongText | None = None

    @model_validator(mode="after")
    def title_cannot_be_null(self) -> Self:
        if "title" in self.model_fields_set and self.title is None:
            raise ValueError("title cannot be null")
        return self


class ContractPatch(_WriteModel):
    customer_company_id: UUID | None = None
    contact_id: UUID | None = None
    product_id: UUID | None = None
    title: Text | None = None
    description: LongText | None = None
    contract_type: ContractType | None = None
    amount: Money | None = None
    contract_date: date | None = None
    ends_on: date | None = None
    warranty_terms: LongText | None = None
    expected_delivery_at: SafeDateTime | None = None
    memo: LongText | None = None

    @model_validator(mode="after")
    def required_fields_cannot_be_null(self) -> Self:
        for field_name in (
            "customer_company_id",
            "product_id",
            "title",
            "contract_type",
            "amount",
            "contract_date",
        ):
            if field_name in self.model_fields_set and getattr(self, field_name) is None:
                raise ValueError(f"{field_name} cannot be null")
        return self


class ContractMove(_WriteModel):
    expected_stage_id: UUID
    stage_id: UUID
    position: Position


class ContractRead(BaseModel):
    id: UUID
    contract_no: str
    customer_company_id: UUID
    customer_company_name: str
    customer_company_region_code: str | None
    contact_id: UUID | None
    contact_name: str | None
    owner_member_id: UUID
    owner_display_name: str
    product_id: UUID | None
    product_name: str | None
    stage_id: UUID
    stage_name: str
    stage_tone: StageTone
    stage_outcome_code: StageOutcome
    stage_position: int
    title: str
    description: str | None
    contract_type: ContractType
    amount: int
    contract_date: date
    ends_on: date | None
    warranty_terms: str | None
    expected_delivery_at: datetime | None
    memo: str | None
    position: int
    created_at: datetime
    updated_at: datetime


class ContractPage(BaseModel):
    items: list[ContractRead]
    skip: int
    limit: int
    total: int
    has_more: bool
    next_skip: int | None


class ContractPageParams(BaseModel):
    model_config = ConfigDict(extra="forbid")

    q: SearchQuery | None = None
    start_date: date | None = None
    end_date: date | None = None
    owner_member_id: list[UUID] | None = None
    stage_id: list[UUID] | None = None
    skip: int = Field(default=0, ge=0, le=9_223_372_036_854_775_807)
    limit: int = Field(default=30, ge=1, le=100)

    @model_validator(mode="after")
    def dates_in_order(self) -> Self:
        if (
            self.start_date is not None
            and self.end_date is not None
            and self.end_date < self.start_date
        ):
            raise ValueError("end_date must not be before start_date")
        return self
